dmipy/core/modeling_framework.py

```diff
@@ -2191,6 +2194,742 @@ def __call__(self, acquisition_scheme, **kwargs):
+            Can be used to recover parameters themselves or other useful
+            functions.
+        """
+        self._check_tissue_model_acquisition_scheme(acquisition_scheme)
+        self._check_model_params_with_acquisition_params(acquisition_scheme)
+        self._check_acquisition_scheme_has_b0s(acquisition_scheme)
+        self._check_if_volume_fractions_are_fixed()
+
+        # estimate S0
+        self.scheme = acquisition_scheme
+        data_ = np.atleast_2d(data)
+        if self.scheme.TE is None or len(np.unique(self.scheme.TE)) == 1:
+            S0 = np.mean(data_[..., self.scheme.b0_mask], axis=-1)
+        else:  # if multiple TE are in the data
+            S0 = np.ones(np.r_[data_.shape[:-1],
+                               len(acquisition_scheme.shell_TE)])
+            for TE_ in self.scheme.shell_TE:
+                TE_mask = self.scheme.shell_TE == TE_
+                TE_mask_shell = self.scheme.TE == TE_
+                TE_b0_mask = np.all([self.scheme.b0_mask, TE_mask_shell],
+                                    axis=0)
+                S0[..., TE_mask] = np.mean(
+                    data_[..., TE_b0_mask], axis=-1)[..., None]
+
+        if mask is None:
+            mask = data_[..., 0] > 0
+        else:
+            mask = np.all([mask, data_[..., 0] > 0], axis=0)
+        mask_pos = np.where(mask)
+
+        N_parameters = len(self.bounds_for_optimization)
+        N_voxels = np.sum(mask)
+
+        # make starting parameters and data the same size
+        x0_ = self.parameter_initial_guess_to_parameter_vector(
+            **self.x0_parameters)
+        x0_ = homogenize_x0_to_data(
+            data_, x0_)
+        x0_bool = np.all(
+            np.isnan(x0_), axis=tuple(np.arange(x0_.ndim - 1)))
+        x0_[..., ~x0_bool] /= self.scales_for_optimization[~x0_bool]
+
+        if use_parallel_processing and not have_pathos:
+            msg = 'Cannot use parallel processing without pathos.'
+            raise ValueError(msg)
+        elif use_parallel_processing and have_pathos:
+            fitted_parameters_lin = [None] * N_voxels
+            if number_of_processors is None:
+                number_of_processors = cpu_count()
+            pool = pp.ProcessPool(number_of_processors)
+            print('Using parallel processing with {} workers.'.format(
+                number_of_processors))
+        else:
+            fitted_parameters_lin = np.empty(
+                np.r_[N_voxels, N_parameters], dtype=float)
+
+        # estimate the spherical mean of the data.
+        data_to_fit = np.zeros(np.r_[data_.shape[:-1], self.scheme.N_shells])
+        for pos in zip(*mask_pos):
+            data_to_fit[pos] = estimate_spherical_mean_multi_shell(
+                data_[pos], self.scheme)
+
+        # TODO: complete the setup of the optimization in model fitting
+        def fit_func(*args, **kwargs):
+            # This function will use the self.forward_model property
```

Comment:
In the SM form the amico forward model matrix would be of shape [N_shells, N_x], not [N_DWI, N_x]